### ISSUE TITLE
Memory corruption leading to Access Violation in .NET 9 using parallel reads on a Trie object

### ISSUE THREAD

**Issue by Treit** [dotnet/runtime#110088]
### Description

A small demonstration program using parallel reads of a Trie data structure lead to a memory corruption issue that crashed the program with an access violation. This does not reproduce on .NET 8 but reproduces on .NET 9.

I constructed a minimal repro project here: https://github.com/treit/ktrieIssue/

The Trie implementation is from this project: https://github.com/kpol/trie/tree/master

### Reproduction Steps

1. Clone this repo: https://github.com/treit/ktrieIssue/
2. dotnet run --framework net9.0

### Expected behavior

No access violation crash

### Actual behavior

The program will eventually crash with an access violation. It may take a minute or two or it may happen immediately in my experience.

```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at KTrie.Trie+<GetDescendantTerminalNodes>d__26.MoveNext()
   at KTrie.Trie+<GetTerminalNodes>d__20.MoveNext()
   at System.Linq.Enumerable+IEnumerableSkipTakeIterator`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at Program+<>c__DisplayClass0_0.<<Main>$>b__0()
   at System.Threading.Tasks.Task`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].InnerInvoke()
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(System.Threading.Thread, System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef,System.Threading.Thread)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool+WorkerThread.WorkerThreadStart()
```

### Regression?

This appears to be a regression because I cannot reproduce it running on .NET 8.

### Configuration

Windows 11 x64 running on .NET 9.

I did not have a chance to try this on Linux. 

I did notice that I could not reproduce it on my local workstation, but it reproduces on both my laptop and my cloud dev box.

**Comment by hez2010** [comment]
Full stacktrace:

```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at System.Collections.Generic.Queue`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Enqueue(System.__Canon)
   at KTrie.Trie+<GetDescendantTerminalNodes>d__26.MoveNext()
   at KTrie.Trie+<GetTerminalNodes>d__20.MoveNext()
   at System.Linq.Enumerable+IEnumerableSelectIterator`2[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Linq.Enumerable+IEnumerableSkipTakeIterator`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at Program+<>c__DisplayClass0_0.<<Main>$>b__0()
   at System.Threading.Tasks.Task`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].InnerInvoke()
   at System.Threading.Tasks.Task+<>c.<.cctor>b__292_0(System.Object)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(System.Threading.Thread, System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef, System.Threading.Thread)
   at System.Threading.Tasks.Task.ExecuteEntryUnsafe(System.Threading.Thread)
   at System.Threading.Tasks.Task.ExecuteFromThreadPool(System.Threading.Thread)
   at System.Threading.ThreadPoolWorkQueue.DispatchWorkItem(System.Object, System.Threading.Thread)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool+WorkerThread.WorkerDoWork(System.Threading.PortableThreadPool, Boolean ByRef)
   at System.Threading.PortableThreadPool+WorkerThread.WorkerThreadStart()
   at System.Threading.Thread+StartHelper.RunWorker()
   at System.Threading.Thread+StartHelper.Run()
   at System.Threading.Thread.StartCallback()
```

Seems that `KTrie` calls `Queue<T>.Enqueue` internally which is not thread-safe, so you should not query `KTrie` in parallel.
Maybe you need to ask the author of `KTrie` to provide a thread-safe version. See https://github.com/kpol/trie/issues/5.

**Comment by jkotas** [comment]
> StelemRef

Dup of #108763 ?

**Comment by AndyAyersMS** [comment]
I wonder if this is a case of calling a managed jit helper in a no gc region.

**Comment by Treit** [comment]
> > StelemRef
> 
> Dup of [#108763](https://github.com/dotnet/runtime/issues/108763) ?

Confirmed that setting `<CETCompat>false</CETCompat>` fixes the issue, per suggestion from hez2010 / skyake on the C# discord.

**Comment by jkotas** [comment]
I have confirmed that it is duplicate of https://github.com/dotnet/runtime/issues/108763 that is scheduled to be fixed in one of the upcoming .NET 9.0 servicing updates #109548.

Thank you for reporting this issue!
